### PR TITLE
[SPARK-31847][CORE][TESTS] DAGSchedulerSuite: Rewrite the test framework to support apply specified spark configurations.

### PR DESCRIPTION
**Pull request by beliefer**
### What changes were proposed in this pull request?
`DAGSchedulerSuite` exists some issue:
`afterEach` and `init` are called when the `SparkConf` of the default `SparkContext` has no configuration that the test case must set. This causes the `SparkContext` initialized in `beforeEach` to be discarded without being used, resulting in waste. On the other hand, the flexibility to add configurations to `SparkConf` should be addressed by the test framework.

Test suites inherits `LocalSparkContext` can be simplified.

### Why are the changes needed?
Reduce overhead about init `SparkContext`.
Rewrite the test framework to support apply specified spark configurations.

### Does this PR introduce _any_ user-facing change?
'No'.

### How was this patch tested?
Jenkins test.

### CODE REVIEW

**Comment by beliefer** [comment]
cc @jiangxb1987 @Ngone51

File: core/src/test/scala/org/apache/spark/SparkConfHelper.scala [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark
+
+trait SparkConfHelper {
+
+  /** Sets all configurations specified in `pairs`, calls `init`, and then calls `testFun` */
+  protected def withSparkConf(pairs: (String, String)*)(testFun: SparkConf => Any): Unit = {
```
Comment (HyukjinKwon) [inline core/src/test/scala/org/apache/spark/SparkConfHelper.scala:22]: `SparkConfHelper.withSparkConf` seems only used `DAGSchedulerSuite`. Let's don't make a trait for now but add it into `DAGSchedulerSuite`.

Comment (beliefer) [reply core/src/test/scala/org/apache/spark/SparkConfHelper.scala:22]: `SparkConfHelper.withSparkConf` will be used in other suite like `TaskSchedulerImplSuite`、`TaskSetManagerSuite`.

**Comment by jiangxb1987** [comment]
It would be really great if you can list the test cases/suites that could get simplified by this change, thanks!

**Comment by beliefer** [comment]
> It would be really great if you can list the test cases/suites that could get simplified by this change, thanks!

I added them to the description of this PR.

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala [hunk 2]
```diff
@@ -295,7 +298,20 @@ class DAGSchedulerSuite extends SparkFunSuite with LocalSparkContext with TimeLi
...
     super.beforeEach()
-    init(new SparkConf())
+  }
+
+  override protected def test(testName: String, testTags: Tag*)(testFun: => Any)
+      (implicit pos: Position): Unit = {
+    testWithSparkConf(testName, testTags: _*)()(testFun)(pos)
+  }
+
+  private def testWithSparkConf(testName: String, testTags: Tag*)
```
Comment (jiangxb1987) [inline core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala:308]: If I understand correctly, you are doing the hack here because you need to modify the SparkConf within `beforeEach` (between `super.beforeEach()` and `init()`). In other words, you don't need to call `testWithSparkConf` instead of `test`, if you don't do extra initialization at the `beforeEach()` stage. Thus, this change is necessarily useful for those test suites you listed, right?

A more staightforward idea would likely to be having a `withConfig(pairs: (String, String)*)` method to create a new SparkConf with the specified configuration values？ The idea doesn't simplify `DAGSchedulerSuite` that much, because you still need to first stop the SparkContext then call `init()` with your own SparkConf, but at least it's not worse than the current approach, and more easy to understand and to be reused.

Comment (beliefer) [reply core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala:308]: For your first question, yes, it is.
For your second question, `withConfig(pairs: (String, String)*)` mean to stop the `SparkContext`  and then call `init()`. So I created the function `testWithSparkConf` which avoid to stop the `SparkContext` then call `init()`.

Comment (Ngone51) [reply core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala:308]: I tend to agree with @jiangxb1987 .  There're only 7 places using the `testWithSparkConf()` comparing to other 81 tests within the `DAGSchedulerSuite`. We could just call `init` inside each test to ease other tests who needs to call `afterEach` inside the `test`.

And I have another idea for the whole thing. That is, we could probably initialize the `SparkContext` like this:


```scala
trait LocalSparkContext ... {
  @transient private var _sc: SparkContext = _
   private val conf = new SparkConf()

  def sc: SparkContext = {
      if (_sc == null) {
        _sc = new SparkContext(conf)
      }
      _sc
  }

  def withConf(pairs: (String, String)*) = {
    if (_sc != null) { 
       // probably, log warning when SparkContext already initialized
       // since configurations won't take effect in this case
     }
     paris.foreach { case (k, v) => conf.set(k, v) }
  }

  override def afterEach(): Unit = {
    try {
      resetSparkContext()
    } finally {
      super.afterEach()
    }
  }

  def resetSparkContext(): Unit = {
    LocalSparkContext.stop(sc)
    ResourceProfile.clearDefaultProfile()
    sc = null
  }
}
```

@beliefer @jiangxb1987 WDYT?

Comment (Ngone51) [reply core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala:308]: I tend to agree with @jiangxb1987 .  There're only 7 places using the `testWithSparkConf()` comparing to other 81 tests within the `DAGSchedulerSuite`. We could just call `init` inside each test to ease other tests who needs to call `afterEach` inside the `test`.

And I have another idea for the whole thing. That is, we could probably initialize the `SparkContext` like this:


```scala
trait LocalSparkContext ... {
  @transient private var _sc: SparkContext = _
   private val conf = new SparkConf()

  def sc: SparkContext = {
      if (_sc == null) {
        _sc = new SparkContext(conf)
      }
      _sc
  }

  def withConf(pairs: (String, String)*) = {
    if (_sc != null) { 
       // probably, log warning when SparkContext already initialized
       // since configurations won't take effect in this case
     }
     paris.foreach { case (k, v) => conf.set(k, v) }
  }

  override def afterEach(): Unit = {
    try {
      resetSparkContext()
    } finally {
      super.afterEach()
    }
  }

  def resetSparkContext(): Unit = {
    LocalSparkContext.stop(sc)
    ResourceProfile.clearDefaultProfile()
    sc = null
  }
}
```

@beliefer @jiangxb1987 WDYT?

Comment (jiangxb1987) [reply core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala:308]: +1 to the above proposal

Comment (beliefer) [reply core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala:308]: OK. I will implement this proposal.

File: core/src/test/scala/org/apache/spark/LocalSparkContext.scala [hunk 3]
```diff
@@ -22,12 +22,37 @@ import org.scalatest.BeforeAndAfterAll
...
+    if (_sc == null) {
+      _sc = new SparkContext(_conf)
+    }
+    _sc
+  }
+
+  def setConf(pairs: (String, String)*): Unit = {
+    if (_sc != null) {
+      logWarning("Because SparkContext already initialized, " +
+        "since configurations won't take effect in this case.")
```
Comment (Ngone51) [inline core/src/test/scala/org/apache/spark/LocalSparkContext.scala:52]: nit: `These configurations ${paris.mkString(", ")} won't take effect since the SparkContext has been already initialized.`

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala [hunk 4]
```diff
@@ -295,11 +295,20 @@ class DAGSchedulerSuite extends SparkFunSuite with LocalSparkContext with TimeLi
 
   override def beforeEach(): Unit = {
     super.beforeEach()
-    init(new SparkConf())
+    firstInit = true
+    setConf("spark.master" -> "local[2]", "spark.app.name" -> "DAGSchedulerSuite")
```
Comment (Ngone51) [inline core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala:299]: We'd better to expose the `conf` via a function to set the conf. `setConf` is designed to be used by the test only.

File: core/src/test/scala/org/apache/spark/LocalSparkContext.scala [hunk 5]
```diff
@@ -22,12 +22,37 @@ import org.scalatest.BeforeAndAfterAll
...
 
+import org.apache.spark.internal.Logging
 import org.apache.spark.resource.ResourceProfile
 
 /** Manages a local `sc` `SparkContext` variable, correctly stopping it after each test. */
-trait LocalSparkContext extends BeforeAndAfterEach with BeforeAndAfterAll { self: Suite =>
+trait LocalSparkContext extends Logging
+  with BeforeAndAfterEach with BeforeAndAfterAll { self: Suite =>
+
+  private var _conf: SparkConf = new SparkConf()
```
Comment (Ngone51) [inline core/src/test/scala/org/apache/spark/LocalSparkContext.scala:32]: I think `SparkConf` should have the default values for `master` and `appName`. So test suites extend it could use the SparkContext directly without any specific configurations when the test doesn't really care.

File: core/src/test/scala/org/apache/spark/LocalSparkContext.scala [hunk 6]
```diff
@@ -22,12 +22,37 @@ import org.scalatest.BeforeAndAfterAll
 import org.scalatest.BeforeAndAfterEach
 import org.scalatest.Suite
 
+import org.apache.spark.internal.Logging
 import org.apache.spark.resource.ResourceProfile
 
 /** Manages a local `sc` `SparkContext` variable, correctly stopping it after each test. */
-trait LocalSparkContext extends BeforeAndAfterEach with BeforeAndAfterAll { self: Suite =>
+trait LocalSparkContext extends Logging
```
[the remark this answers is not in the document]
Comment (beliefer) [reply core/src/test/scala/org/apache/spark/LocalSparkContext.scala:29]: OK

File: core/src/test/scala/org/apache/spark/LocalSparkContext.scala (hunk 3 above)
Comment (beliefer) [reply core/src/test/scala/org/apache/spark/LocalSparkContext.scala:52]: OK

File: core/src/test/scala/org/apache/spark/LocalSparkContext.scala (hunk 5 above)
Comment (beliefer) [reply core/src/test/scala/org/apache/spark/LocalSparkContext.scala:32]: OK

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala (hunk 4 above)
Comment (beliefer) [reply core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala]: OK

**Comment by beliefer** [comment]
retest this please

File: core/src/test/scala/org/apache/spark/LocalSparkContext.scala [hunk 7]
```diff
@@ -22,12 +22,39 @@ import org.scalatest.BeforeAndAfterAll
 import org.scalatest.BeforeAndAfterEach
 import org.scalatest.Suite
 
+import org.apache.spark.internal.Logging
 import org.apache.spark.resource.ResourceProfile
 
 /** Manages a local `sc` `SparkContext` variable, correctly stopping it after each test. */
-trait LocalSparkContext extends BeforeAndAfterEach with BeforeAndAfterAll { self: Suite =>
+trait LocalSparkContext extends BeforeAndAfterEach
+  with BeforeAndAfterAll with Logging{ self: Suite =>
```
Comment (Ngone51) [inline core/src/test/scala/org/apache/spark/LocalSparkContext.scala:30]: nit: needs a space before `{`

File: core/src/test/scala/org/apache/spark/LocalSparkContext.scala [hunk 8]
```diff
@@ -22,12 +22,39 @@ import org.scalatest.BeforeAndAfterAll
...
+
+  private var _conf: SparkConf = defaultSparkConf
 
   @transient var sc: SparkContext = _
+  @transient private var _sc: SparkContext = _
+
+  def getSparkConf: SparkConf = _conf
```
Comment (Ngone51) [inline core/src/test/scala/org/apache/spark/LocalSparkContext.scala:37]: nit: `def conf`?

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala [hunk 9]
```diff
@@ -235,7 +235,7 @@ class DAGSchedulerSuite extends SparkFunSuite with LocalSparkContext with TimeLi
       _endedTasks.toSet
     }
 
-    private def waitForListeners(): Unit = sc.listenerBus.waitUntilEmpty()
+    private def waitForListeners(): Unit = sparkCtx.listenerBus.waitUntilEmpty()
```
Comment (Ngone51) [inline core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala:238]: I'm thinking if we could have a new copy of `LocalSparkContext`, e.g., `LocalSC`, then, we don't need to change from `sc` to `sparkCtx`? And for other test suites, we only need to replace `LocalSparkContext` with `LocalSC`?

File: core/src/test/scala/org/apache/spark/LocalSparkContext.scala (hunk 7 above)
Comment (beliefer) [reply core/src/test/scala/org/apache/spark/LocalSparkContext.scala:30]: OK

File: core/src/test/scala/org/apache/spark/LocalSparkContext.scala [hunk 10]
```diff
@@ -22,12 +22,39 @@ import org.scalatest.BeforeAndAfterAll
...
+  def sparkCtx: SparkContext = {
+    if (_sc == null) {
+      _sc = new SparkContext(_conf)
+    }
+    _sc
+  }
+
+  def setConf(pairs: (String, String)*): Unit = {
+    if (_sc != null) {
+      logWarning(s"These configurations ${pairs.mkString(", ")} won't take effect" +
```
Comment (Ngone51) [inline core/src/test/scala/org/apache/spark/LocalSparkContext.scala:53]: nit: need a space after "effect"

File: core/src/test/scala/org/apache/spark/LocalSparkContext.scala (hunk 8 above)
Comment (beliefer) [reply core/src/test/scala/org/apache/spark/LocalSparkContext.scala:37]:
```
[error] /home/jenkins/workspace/SparkPullRequestBuilder@2/resource-managers/mesos/src/test/scala/org/apache/spark/scheduler/cluster/mesos/MesosCoarseGrainedSchedulerBackendSuite.scala:49: overriding method sparkConf in trait LocalSparkContext of type => org.apache.spark.SparkConf;
[error]  variable sparkConf has weaker access privileges; it should not be private
[error]   private var sparkConf: SparkConf = _
```

File: core/src/test/scala/org/apache/spark/LocalSparkContext.scala (hunk 10 above)
Comment (beliefer) [reply core/src/test/scala/org/apache/spark/LocalSparkContext.scala:53]: OK

File: core/src/test/scala/org/apache/spark/LocalSparkContext.scala [hunk 11]
```diff
@@ -22,12 +22,39 @@ import org.scalatest.BeforeAndAfterAll
...
+   * renamed to `sc` and the variable `sc` will be deleted.
+   */
+  def sparkCtx: SparkContext = {
+    if (_sc == null) {
+      _sc = new SparkContext(_conf)
+    }
+    _sc
+  }
+
+  def setConf(pairs: (String, String)*): Unit = {
```
Comment (Ngone51) [inline core/src/test/scala/org/apache/spark/LocalSparkContext.scala:51]: `setConf` just sounds like an attribute setting function which needs to invoke like `conf.setConf`. After we exposing the `conf`, I think it's more straightforward to use it like `conf.setConf(...)` or `conf.setAll(...)`.  Here, I think what we want is a helper method, e.g., `withConf`. And `withXXX` is more like a common pattern in Spark. We can use it to work with other helper methods like, `withTempDir`, `withListener`, and others in the future.

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala (hunk 9 above)
Comment (beliefer) [reply core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala]: OK

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala [hunk 12]
```diff
@@ -460,7 +470,8 @@ class DAGSchedulerSuite extends SparkFunSuite with LocalSparkContext with TimeLi
     val shuffleDepB = new ShuffleDependency(rddB, new HashPartitioner(1))
     val s_B = shuffleDepB.shuffleId
 
-    val rddC = new MyRDD(sc, 1, List(shuffleDepA, shuffleDepB), tracker = mapOutputTracker)
+    val rddC =
+      new MyRDD(sc, 1, List(shuffleDepA, shuffleDepB), tracker = mapOutputTracker)
```
Comment (Ngone51) [inline core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala:474]: unnecessary change?

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala [hunk 13]
```diff
@@ -1104,7 +1104,8 @@ class DAGSchedulerSuite extends SparkFunSuite with LocalSparkContext with TimeLi
 
     val shuffleOneRdd = new MyRDD(sc, 2, Nil).cache()
     val shuffleDepOne = new ShuffleDependency(shuffleOneRdd, new HashPartitioner(2))
-    val shuffleTwoRdd = new MyRDD(sc, 2, List(shuffleDepOne), tracker = mapOutputTracker).cache()
+    val shuffleTwoRdd =
+      new MyRDD(sc, 2, List(shuffleDepOne), tracker = mapOutputTracker).cache()
```
Comment (Ngone51) [inline core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala:1108]: unnecessary change?

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala [hunk 14]
```diff
@@ -1153,7 +1154,8 @@ class DAGSchedulerSuite extends SparkFunSuite with LocalSparkContext with TimeLi
...
```
Comment (Ngone51) [inline core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala:1158]: ditto

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala [hunk 15]
```diff
@@ -1757,7 +1759,8 @@ class DAGSchedulerSuite extends SparkFunSuite with LocalSparkContext with TimeLi
     val shuffleDep2 = new ShuffleDependency(shuffleMapRdd2, new HashPartitioner(2))
 
     val reduceRdd1 = new MyRDD(sc, 2, List(shuffleDep1), tracker = mapOutputTracker)
-    val reduceRdd2 = new MyRDD(sc, 2, List(shuffleDep1, shuffleDep2), tracker = mapOutputTracker)
+    val reduceRdd2 =
+      new MyRDD(sc, 2, List(shuffleDep1, shuffleDep2), tracker = mapOutputTracker)
```
Comment (Ngone51) [inline core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala:1763]: ditto

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala [hunk 16]
```diff
@@ -1954,7 +1957,8 @@ class DAGSchedulerSuite extends SparkFunSuite with LocalSparkContext with TimeLi
   test("cached post-shuffle") {
     val shuffleOneRdd = new MyRDD(sc, 2, Nil).cache()
     val shuffleDepOne = new ShuffleDependency(shuffleOneRdd, new HashPartitioner(2))
-    val shuffleTwoRdd = new MyRDD(sc, 2, List(shuffleDepOne), tracker = mapOutputTracker).cache()
+    val shuffleTwoRdd =
+      new MyRDD(sc, 2, List(shuffleDepOne), tracker = mapOutputTracker).cache()
```
Comment (Ngone51) [inline core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala:1961]: ditto

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala [hunk 17]
```diff
@@ -2255,7 +2259,8 @@ class DAGSchedulerSuite extends SparkFunSuite with LocalSparkContext with TimeLi
     val shuffleDep = new ShuffleDependency(rdd1, new HashPartitioner(1))
     val narrowDep = new OneToOneDependency(rdd2)
     val shuffleId = shuffleDep.shuffleId
-    val reduceRdd = new MyRDD(sc, 1, List(shuffleDep, narrowDep), tracker = mapOutputTracker)
+    val reduceRdd =
+      new MyRDD(sc, 1, List(shuffleDep, narrowDep), tracker = mapOutputTracker)
```
Comment (Ngone51) [inline core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala:2263]: ditto

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala [hunk 18]
```diff
@@ -2851,11 +2856,13 @@ class DAGSchedulerSuite extends SparkFunSuite with LocalSparkContext with TimeLi
 
     val shuffleDep1 = new ShuffleDependency(shuffleMapRdd1, new HashPartitioner(2))
     val shuffleId1 = shuffleDep1.shuffleId
-    val shuffleMapRdd2 = new MyRDD(sc, 2, List(shuffleDep1), tracker = mapOutputTracker)
+    val shuffleMapRdd2 =
+      new MyRDD(sc, 2, List(shuffleDep1), tracker = mapOutputTracker)
```
Comment (Ngone51) [inline core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala:2860]: ditto

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala [hunk 19]
```diff
@@ -2851,11 +2856,13 @@ class DAGSchedulerSuite extends SparkFunSuite with LocalSparkContext with TimeLi
...
 
     val shuffleDep2 = new ShuffleDependency(shuffleMapRdd2, new HashPartitioner(2))
     val shuffleId2 = shuffleDep2.shuffleId
-    val finalRdd = new MyRDD(sc, 2, List(shuffleDep2), tracker = mapOutputTracker)
+    val finalRdd =
+      new MyRDD(sc, 2, List(shuffleDep2), tracker = mapOutputTracker)
```
Comment (Ngone51) [inline core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala:2865]: ditto

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala [hunk 20]
```diff
@@ -3198,7 +3201,8 @@ class DAGSchedulerSuite extends SparkFunSuite with LocalSparkContext with TimeLi
     val treqs2 = new TaskResourceRequests().cpus(2)
     val rp2 = new ResourceProfileBuilder().require(ereqs2).require(treqs2).build
 
-    val rdd = sc.parallelize(1 to 10).withResources(rp1).map(x => (x, x)).withResources(rp2)
+    val rdd =
+      sc.parallelize(1 to 10).withResources(rp1).map(x => (x, x)).withResources(rp2)
```
Comment (Ngone51) [inline core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala:3205]: ditto

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala [hunk 21]
```diff
@@ -3221,18 +3222,16 @@ class DAGSchedulerSuite extends SparkFunSuite with LocalSparkContext with TimeLi
...
```
Comment (Ngone51) [inline core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala:3226]: ditto

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala [hunk 22]
```diff
@@ -3242,14 +3241,16 @@ class DAGSchedulerSuite extends SparkFunSuite with LocalSparkContext with TimeLi
     val treqs2 = new TaskResourceRequests().cpus(2)
     val rp2 = new ResourceProfileBuilder().require(ereqs2).require(treqs2).build
 
-    val rdd = sc.parallelize(1 to 10).withResources(rp1).map(x => (x, x)).withResources(rp2)
+    val rdd = sc.parallelize(1 to 10)
+      .withResources(rp1).map(x => (x, x)).withResources(rp2)
```
Comment (Ngone51) [inline core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala:3245]: ditto

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala [hunk 23]
```diff
@@ -3242,14 +3241,16 @@ class DAGSchedulerSuite extends SparkFunSuite with LocalSparkContext with TimeLi
...
     val (_, resourceprofiles) = scheduler.getShuffleDependenciesAndResourceProfiles(rdd)
     val mergedRp = scheduler.mergeResourceProfilesForStage(resourceprofiles)
     assert(mergedRp.getTaskCpus.get == 2)
     assert(mergedRp.getExecutorCores.get == 4)
 
     // test that instead of creating a new merged profile, we use the already created one
-    val rdd2 = sc.parallelize(1 to 10).withResources(rp1).map(x => (x, x)).withResources(rp2)
+    val rdd2 =
+      sc.parallelize(1 to 10).withResources(rp1).map(x => (x, x)).withResources(rp2)
```
Comment (Ngone51) [inline core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala:3253]: ditto

File: core/src/test/scala/org/apache/spark/LocalSC.scala [hunk 24]
```diff
@@ -0,0 +1,103 @@
...
+import _root_.io.netty.util.internal.logging.{InternalLoggerFactory, Slf4JLoggerFactory}
+import org.scalatest.BeforeAndAfterAll
+import org.scalatest.BeforeAndAfterEach
+import org.scalatest.Suite
+
+import org.apache.spark.internal.Logging
+import org.apache.spark.resource.ResourceProfile
+
+/** Manages a local `sc` `SparkContext` variable, correctly stopping it after each test. */
+trait LocalSC extends BeforeAndAfterEach
```
Comment (Ngone51) [inline core/src/test/scala/org/apache/spark/LocalSC.scala:29]: Could explain more about why we need this?

File: core/src/test/scala/org/apache/spark/LocalSC.scala [hunk 25]
```diff
@@ -0,0 +1,103 @@
...
+  with BeforeAndAfterAll with Logging { self: Suite =>
+
+  private var _conf: SparkConf = defaultSparkConf
+
+  @transient private var _sc: SparkContext = _
+
+  def getSparkConf: SparkConf = _conf
```
Comment (Ngone51) [inline core/src/test/scala/org/apache/spark/LocalSC.scala:36]: Would `def conf: SparkConf` still cause the compile error after we moving all the things to a separate file?

File: core/src/test/scala/org/apache/spark/LocalSC.scala [hunk 26]
```diff
@@ -0,0 +1,103 @@
...
+   * renamed to `sc` and the variable `sc` will be deleted.
+   */
+  def sc: SparkContext = {
+    if (_sc == null) {
+      _sc = new SparkContext(_conf)
+    }
+    _sc
+  }
+
+  def setConf(pairs: (String, String)*): Unit = {
```
Comment (Ngone51) [inline core/src/test/scala/org/apache/spark/LocalSC.scala:50]: In case you miss this comment: https://github.com/apache/spark/pull/29228#discussion_r474396405

File: core/src/test/scala/org/apache/spark/LocalSC.scala [hunk 27]
```diff
@@ -0,0 +1,103 @@
...
+  def resetSparkContext(): Unit = {
+    LocalSC.stop(_sc)
+    ResourceProfile.clearDefaultProfile()
+    _sc = null
+    _conf = defaultSparkConf
+  }
+
+  private def defaultSparkConf: SparkConf = new SparkConf()
+    .setMaster("local[2]").setAppName(s"${this.getClass.getSimpleName}")
+
```
Comment (Ngone51) [inline core/src/test/scala/org/apache/spark/LocalSC.scala:80]: nit: no blank line.

File: core/src/test/scala/org/apache/spark/LocalSC.scala [hunk 28]
```diff
@@ -0,0 +1,103 @@
...
+  /** Runs `f` by passing in `sc` and ensures that `sc` is stopped. */
+  def withSpark[T](sc: SparkContext)(f: SparkContext => T): T = {
+    try {
+      f(sc)
+    } finally {
+      stop(sc)
+    }
+  }
+
```
Comment (Ngone51) [inline core/src/test/scala/org/apache/spark/LocalSC.scala:100]: nit: no blank line.

File: core/src/test/scala/org/apache/spark/LocalSparkContext.scala (hunk 11 above)
Comment (beliefer) [reply core/src/test/scala/org/apache/spark/LocalSparkContext.scala]: OK

File: core/src/test/scala/org/apache/spark/LocalSC.scala (hunk 25 above)
Comment (beliefer) [reply core/src/test/scala/org/apache/spark/LocalSC.scala:36]: No

File: core/src/test/scala/org/apache/spark/LocalSC.scala (hunk 26 above)
Comment (beliefer) [reply core/src/test/scala/org/apache/spark/LocalSC.scala:50]: OK

File: core/src/test/scala/org/apache/spark/LocalSC.scala (hunk 25 above)
Comment (Ngone51) [reply core/src/test/scala/org/apache/spark/LocalSC.scala:36]: Then, let's use `def conf: SparkConf`.

File: core/src/test/scala/org/apache/spark/LocalSC.scala (hunk 24 above)
Comment (beliefer) [reply core/src/test/scala/org/apache/spark/LocalSC.scala:37]: OK

File: core/src/test/scala/org/apache/spark/LocalSC.scala (hunk 25 above)
Comment (beliefer) [reply core/src/test/scala/org/apache/spark/LocalSC.scala]: OK

**Comment by beliefer** [comment]
retest this please

**Comment by beliefer** [comment]
retest this please

**Comment by beliefer** [comment]
retest this please

**Comment by Ngone51** [comment]
LGTM. BTW, could you update the PR description? You may not need to list all the test suites with the latest change but just say test suites inherits `LocalSparkContext` can be simplified.

@jiangxb1987 Could you also take a look?

**Comment by beliefer** [comment]
@Ngone51 I updated the description of this PR.

File: core/src/test/scala/org/apache/spark/LocalSC.scala [hunk 29]
```diff
@@ -0,0 +1,99 @@
...
+/**
+ * Manages a local `sc` `SparkContext` variable, correctly stopping it after each test.
+ *
+ * Note: this class is a copy of [[LocalSparkContext]]. Why copy it? Reduce conflict. Because
+ * many test suites use [[LocalSparkContext]] and overwrite some variable or function (e.g.
+ * sc of LocalSparkContext), there occurs conflict when we refactor the `sc` as a new function.
+ * After migrating all test suites that use [[LocalSparkContext]] to use [[LocalSC]], we will
+ * delete the original [[LocalSparkContext]] and rename [[LocalSC]] to [[LocalSparkContext]].
+ */
+trait LocalSC extends BeforeAndAfterEach
```
Comment (jiangxb1987) [inline core/src/test/scala/org/apache/spark/LocalSC.scala:37]: Since this class is only used for temporary purpose, can we name it as `TempLocalSparkContext` ? TBH I don't like the `SC` name which is very vague to me.

**Comment by jiangxb1987** [comment]
LGTM otherwise

File: core/src/test/scala/org/apache/spark/LocalSC.scala (hunk 29 above)
Comment (beliefer) [reply core/src/test/scala/org/apache/spark/LocalSC.scala:37]: OK

**Review by jiangxb1987 — APPROVED** [review]
LGTM

**Comment by jiangxb1987** [comment]
retest this please

**Comment by beliefer** [comment]
retest this please

**Comment by beliefer** [comment]
retest this please

**Comment by beliefer** [comment]
retest this please

**Comment by beliefer** [comment]
cc @jiangxb1987

**Comment by jiangxb1987** [comment]
retest this please

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.
